### PR TITLE
move batching to preprocessor to prepare for seq parallel

### PR DESCRIPTION
**Pull request by AlexPiche**
Screenshot 2025-07-04 at 3 58 43 PM

Screenshot 2025-07-12 at 12 34 10 PM

1. Batch Creation in Preprocessor (pipelinerl/preprocess.py)

  - Moved collation logic (collate and collate_packed) from fine-tuning to preprocessing stage
  - Preprocessor now outputs BatchEncoding objects instead of individual samples
  - Supports both standard batching and sequence packing modes
  - Implements per-worker sample tracking to ensure balanced data distribution

  2. Ring Buffer Implementation

  - Replaced list-based buffer with a deque-based ring buffer
  - Prevents unbounded memory growth when trainer is slower than preprocessor
  - Automatically discards old data when buffer is full, maintaining a sliding window of recent samples

  3. Trainer-Preprocessor Synchronization

  - Added bidirectional communication between trainer and preprocessor
  - Preprocessor reads trainer_status messages to track how many samples have been processed
  - Only sends new batches when trainer is ready (based on target_published_samples)
  - Prevents overwhelming the trainer with excessive queued data

  4. State Persistence

  - Added preprocessor_state.json to persist published samples count across restarts
  - Loads trainer's last processed sample count from training checkpoint
  - Ensures continuity when resuming training

  5. Debug and Monitoring Improvements

  - Added detailed per-worker sample tracking metrics
  - Enhanced logging for batch creation and worker synchronization
  - Better error handling with explicit re-raising for debugging

  Technical Details

  - Supports both standard batching and sequence packing modes
  - Maintains backward compatibility with existing training loop
  - Per-worker sample tracking ensures fair distribution across training GPUs
  - Thread-safe status reading using dedicated worker thread

Tests

- [x] `seq_packing=true` and `seq_packing=false` gives the same loss.
- [ ] Loss matches main.

### CODE REVIEW

File: pipelinerl/preprocess.py [hunk 1]
```diff
@@ -346,23 +384,58 @@ def run_preprocessing_loop(
...
 
-    published_samples = 0
+    # Load published samples from state file
+    state_path = preprocess_dir / "preprocessor_state.json"
+    
+    # Handle force_restart flag
+    if cfg.force_restart and state_path.exists():
+        logger.info(f"force_restart=True, removing existing preprocessor state at {state_path}")
+        state_path.unlink()
+    published_samples = load_preprocessor_state(state_path)
```
[the remark this answers is not in the document]
Comment (rizar) [reply pipelinerl/preprocess.py:437]: What you can do here instead is that the preprocessor can read an "init" message from the trainer and count the samples from the number of samples consumed by the trainer.

File: pipelinerl/streams.py [hunk 2]
```diff
@@ -255,7 +255,17 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+            # Convert Pydantic model to dict, handling tensors
+            data_dict = data.model_dump()
+            # Convert any tensors to numpy arrays for JSON serialization
+            for key, value in data_dict.items():
+                if hasattr(value, 'numpy'):
+                    data_dict[key] = value.numpy()
+                elif hasattr(value, 'tolist'):
+                    data_dict[key] = value.tolist()
+            data = data_dict
+        elif hasattr(data, '__class__') and 'PipelineBatchEncoding' in str(type(data)):
```
Comment (rizar) [inline pipelinerl/streams.py:267]: why not `isinstance(data, PipelineBatchEncoding)`?

**Review by rizar — COMMENTED** [review]
Good work in progress! I have left some comments in the code.
